docs/user-docs/build_grub.md

```diff
@@ -0,0 +1,210 @@
+
+```
+docker run --rm -it -v "$PWD:$PWD" -w "$PWD" \
+           -e HOME="$PWD/home" \
+           --user "$(id -u):$(id -g)" \
+           ghcr.io/trenchboot/trenchboot-sdk:master /bin/bash
+```
+
+Setting `$HOME` is necessary because GRUB will try to use `ccache` which fails
+if there is no `$HOME`.  To install GRUB for legacy boot, the container might
+need to be started as `root` (inside of the container, don't prepend `sudo` to
+the command):
+
+```
+docker run --rm -it -v "$PWD:$PWD" -w "$PWD" \
+           --user root:root \
+           ghcr.io/trenchboot/trenchboot-sdk:master /bin/bash
+```
+
+[trenchboot-sdk]: https://github.com/TrenchBoot/trenchboot-sdk
+
+## EFI version
+
+### Build
+
+```bash
+# check out the sources
+git clone --branch grub-sl-2.12-v9 --depth 1 \
+          https://github.com/TrenchBoot/grub.git
+cd grub
+
+# make the checkout buildable
+./bootstrap
+
+# configure
+mkdir build
+cd build
+../configure --with-platform=efi --target=x86_64
+
+# build
+make -j$(nproc)
+```
+
+### Setup
+
+```bash
+# also from inside build/ directory
+echo 'configfile ${cmdpath}/grub.cfg' > embedded.cfg
+./grub-mkimage -O x86_64-efi -o grubx64.efi -p / -c embedded.cfg -d grub-core \
+               all_video boot btrfs cat chain configfile echo efifwsetup \
+               efinet ext2 fat font gfxmenu gfxterm gzio halt hfsplus iso9660 \
+               jpeg loadenv loopback lvm mdraid09 mdraid1x minicmd normal \
+               part_apple part_msdos part_gpt password_pbkdf2 png reboot \
+               regexp search search_fs_uuid search_fs_file search_label serial \
+               sleep syslinuxcfg test tftp video xfs backtrace http linux usb \
+               usbserial_common usbserial_pl2303 usbserial_ftdi \
+               usbserial_usbdebug keylayouts at_keyboard multiboot2
+```
+
+The first command creates an embedded config which sources grub.cfg in the
+directory in which GRUB image is located.
+
+The second command produces the UEFI GRUB image named `grubx64.efi`.  The image
+along with a suitable `grub.cfg` needs to be installed on a EFI System
+Partition (ESP) and pointed at so firmware can find it (via Setup UI or a tool
+like `efibootmgr`).  It can also be started manually from a UEFI shell for a
+test.
+
+## Legacy version
+
+### Build
+
+```bash
+# check out the sources
+git clone --branch tb-2.12-57-v1 --depth 1 https://github.com/TrenchBoot/grub.git
+cd grub
+
+# make the checkout buildable
+./bootstrap
+
+# configure
+mkdir build
+cd build
+../configure --prefix=$PWD/local-install --target=i386
+
+# build
+make -j$(nproc)
+```
+
+### Setup
+
+```bash
+# also from inside build/ directory
+make install
+local-install/sbin/grub-install --target=i386-pc \
+                                --boot-directory PATH-TO-BOOT \
+                                DISK-DEVICE
+```
+
+In the second command above replace:
+
+- `PATH-TO-BOOT` with the path to where the boot partition is mounted (`grub/`
```

Comment:
The mounted partition and `/dev/...` file need to be mounted to docker to use them here. Maybe add it in some way to the docker command at the top of the file?

Reply:
Mounting them is not enough. `grub-install` says 
```
local-install/sbin/grub-install: warning: disk does not exist, so falling back to partition device /dev/sdc1.
local-install/sbin/grub-install: error: disk `hostdisk//dev/sdc1' not found.
```
I am not sure what this `hostdisk/` means there, but `/dev/sdc` and `/dev/sdc1` do exist in the containers filesystem.

Reply:
Interestingly, the `grub-install` does seem to be able to access the device, because it creates a `grub/` directory, but nothing else.

Reply:
Right, I don't think I've tested installation from Docker, only the build.  See https://github.com/TrenchBoot/TrenchBoot.github.io/compare/948f90c734c3d89b117c0b13b37014064819b8a8..892edc3ec0dccbe1c9fc7564c5aa3f2f15f4f9e3

That nonsensical error is caused by GRUB not being able to write to device files, so you probably weren't `root` in the container.

Reply:
I have been copying the commands directly from the documentation and the command prompt did say the user was root so that should not be the cause of this error?

Reply:
The `--privileged` flag solves the problem with privileges.

Reply:
> I have been copying the commands directly from the documentation and the command prompt did say the user was root so that should not be the cause of this error?

Then the device probably just didn't exist without the `--privileged`  flag.